ncd-calculator/src/services/kgrid.ts

```diff
@@ -0,0 +1,595 @@
+            uniqueIdsMap.set(object.id, object);
+        }
+    })
+
+    const uniqueObjects = Array.from(uniqueIdsMap.values());
+    if (uniqueObjects.length < width * height) {
+        throw new Error(`Not enough unique objects (${uniqueObjects.length}) for a ${width}×${height} grid`);
+    }
+
+    const shuffledObjects = [...uniqueObjects];
+    for (let i = shuffledObjects.length - 1; i >= 0; i--) {
+        const j = Math.floor(Math.random() * (i + 1));
+        [shuffledObjects[i], shuffledObjects[j]] = [shuffledObjects[j], shuffledObjects[i]];
+    }
+
+    const selected = shuffledObjects.slice(0, width * height);
+    const grid: string[][] = [];
+
+    let index = 0;
+    for (let i = 0; i < height; i++) {
+        grid[i] = [];
+        for (let j = 0; j < width; j++) {
+            grid[i][j] = selected[index].id;
+            index++;
+        }
+    }
+    const rows = grid.map((row, i) => {
+        return `Row ${i}:` + row.join(", ");
+    });
+    console.log("Created initial grid with IDs: " + rows);
+
+
+    const ncdMatrix = precomputeNCDMatrix(objects);
+
+    const gridState = {
+        width, height, grid, ncdMatrix, objectiveValue: 0
+    }
+
+    gridState.objectiveValue = calculateObjective(gridState);
+    return ensureStringIds(gridState);
+}
+
+
+export const calculateMatchPercentage = (grid1: GridState, grid2: GridState): number => {
+    if (grid1.width !== grid2.width || grid1.height !== grid2.height) {
+        throw new Error("Grids must have the same dimensions to calculate match percentage");
+    }
+
+    let matches = 0;
+    const totalCells = grid1.width * grid1.height;
+
+    for (let i = 0; i < grid1.height; i++) {
+        for (let j = 0; j < grid1.width; j++) {
+            if (grid1.grid[i][j] === grid2.grid[i][j]) {
+                matches++;
+            }
+        }
+    }
+
+    return matches / totalCells;
+}
+
+export const selectRandomCell = (grid: GridState): Position => {
+    const i = Math.floor(Math.random() * grid.height);
+    const j = Math.floor(Math.random() * grid.width);
+
+    console.log(`Selected random cell at position (${i},${j})`);
+    return {i, j};
+}
+
+
+export const swapCells = (grid: GridState, pos1: Position, pos2: Position): GridState => {
+    console.log(`Swapping cells: (${pos1.i},${pos1.j}) with (${pos2.i},${pos2.j})`);
+
+    const newGrid = deepCopy(grid);
+    const value1 = String(grid.grid[pos1.i][pos1.j]);
+    const value2 = String(grid.grid[pos2.i][pos2.j]);
+
+    newGrid.grid[pos1.i][pos1.j] = value2;
+    newGrid.grid[pos2.i][pos2.j] = value1;
+
+    console.log(`Swapped ${value1} at (${pos1.i},${pos1.j}) with ${value2} at (${pos2.i},${pos2.j})`);
+    newGrid.objectiveValue = calculateObjective(newGrid);
+    return ensureStringIds(newGrid);
+}
+
+
+const OPTIMIZE_STEP_OPTIONS = {
+    initialTemperature: 10.0,  // Starting temperature (higher = more exploration)
+    finalTemperature: 0.01,    // Final temperature
```

Comment:
i think final temperature should be 0
also we don't need resetProbability then i suppose
it never needs to reset